casper/src/main/scala/coop/rchain/casper/blocks/BlockReceiver.scala

```diff
@@ -243,25 +253,32 @@ object BlockReceiver {
           val markReceivedAndStore =
             for {
               // Save block to block store, resolve parents to request
-              _ <- BlockStore[F].put(block)
+              blockStored <- BlockStore[F].contains(block.blockHash)
+              _ <- (BlockStore[F].put(block) *> BlockRetriever[F].ackReceived(block.blockHash))
+                    .whenA(!blockStored)
+
               parents <- block.justifications
                           .traverse { hash =>
                             BlockStore[F].contains(hash).not.map((hash, _))
                           }
               pendingRequests <- state.modify(_.endStored(block.blockHash, parents))
 
-              // Notify BlockRetriever of finished validation of block
-              _ <- BlockRetriever[F].ackReceived(block.blockHash)
-
-              // Send request for missing dependencies
-              _ <- requestMissingDependencies(pendingRequests)
-
               // Check if block have all dependencies in the DAG
-              dag <- BlockDagStorage[F].getRepresentation
-              hasAllDeps = pendingRequests.isEmpty &&
-                block.justifications.forall(dag.contains)
-
-              _ <- receiverOutputQueue.enqueue1(block.blockHash).whenA(hasAllDeps)
+              dag        <- BlockDagStorage[F].getRepresentation
+              hasAllDeps = block.justifications.forall(dag.contains)
+
+              // If replay was interrupted, block was stored but not validated or added to the DAG.
+              // Validation needs to be restarted for such blocks
+              parentsToValidate <- block.justifications.filterA(toValidate)
+
+              _ <- if (hasAllDeps) {
+                    receiverOutputQueue.enqueue1(block.blockHash)
+                  } else {
+                    requestMissingDependencies(pendingRequests).whenA(pendingRequests.nonEmpty) *>
+                      sendToValidate(parentsToValidate, receiverOutputQueue).whenA(
+                        parentsToValidate.nonEmpty
+                      )
+                  }
```

Comment:
What this change is fixing? The issue was that node didn't start validating after restart for which you have fix in NodeRunning.

Reply:
Consider situation when `#1` received/stored but not validated. Dependencies for `#2` will not be resolved. In the same time `pendingRequests` for `#2` will be empty because `#1` received/stored. For the parents which was stored but not validated we should put it to validation, i.e. continue process which was interrupted with node stopping. In the last version of code I'm putting such parents directly to the BlockReceiver input queue.

![image](https://user-images.githubusercontent.com/1443855/185411455-70eb215d-da83-4375-a334-50b03887f988.png)

Reply:
In general case you can have 1000 ancestors of the block that are not validated yet (this this picture with lots #1) - and this code on receipt of the hash of this very tip will trigger exchange of all these hashes down till the one which can be validated. Would be good to just have an ability to start validating where node stopped. And never ask for any block which is already in block store.

Reply:
@nzpr
> Would be good to just have an ability to start validating where node stopped

I'm not sure how to detect where node stopped. We have BlockStore for received blocks and DAG for validated blocks. So we can calculate difference and validate blocks which presented in BS but not in BDS. But such blocks unordered and it might be problematic to validate it in this case. Additionally, BlockReceiver state is not saved to persistent storage.

> And never ask for any block which is already in block store

In the current solution, only unaccepted blocks are requested.